src/librustc_mir/const_eval.rs

```diff
@@ -365,13 +365,19 @@ impl<'a, 'mir, 'tcx> interpret::Machine<'a, 'mir, 'tcx>
         ret: Option<mir::BasicBlock>,
     ) -> EvalResult<'tcx, Option<&'mir mir::Mir<'tcx>>> {
         debug!("eval_fn_call: {:?}", instance);
-        if !ecx.tcx.is_const_fn(instance.def_id()) {
+        // Execution might have wandered off into other crates, so we cannot to a stability-
+        // sensitive check here.  But we can at least rule out functions that are not const
+        // at all.
+        if !ecx.tcx.is_const_fn_raw(instance.def_id()) {
```

Comment:
> There are several feature gates though, not just const_fn. Maybe it should just check for being on a stable compiler?

There's no such check. Using the `const_fn` feature gate as a strawman should be fine imo.

Reply:
There is `is_nightly_build`. But that's not great, it'd mean we would only catch problems when branching off a beta.

I will try the `const_fn` gate, let's see if that works. I think it might have to be `const_fn || const_transmute || const_[...]` though.

Reply:
> const qualification is a static check. It cannot detect what you do once you start doing unsafe things. Current const qualification successfully prevents this from being a problem. Once we allow arbitrary transmuting, const qualification can't help us anymore.

Sure, but I mean in terms of reporting feature gates to add -- you shouldn't add a feature gate to get around a bad `transmute`...

Reply:
I don't care about bad transmutes. I just want a way for the regular unstable code to work, but that will also allow bad transmutes.

Reply:
> Using the const_fn feature gate as a strawman should be fine imo.

It doesn't work, `consts/const-eval/strlen.rs` fails because it does not have the `const_fn` feature enabled. The feature it has enabled is a library feature so I cannot even reasonably test for it.

I think `is_nightly_build` is the only option, if we want to do something stronger than what we do currently.

Reply:
That's fine. That's why I wrote 

> extend the error message to contain a hint about adding the feature gate

Reply:
But that's exactly what I have been talking about above. :)  Adding a feature gate can only ever help for strange transmuting code.

Or are you saying `strlen.rs` *should* add the `const_fn` feature gate? I do not think so, the entire point is to be able to opt into a few bits of `const fn` without opting into all of it.

Reply:
We can also add an `transitive_unstable_const_fn_calls` feature gate

Reply:
I'm fine either way. If someone uses undefined behavior it to circumvent stability, that's not our problem and we may break their code.

Reply:
I sympathize with that view, but I'd like to have this check as an additional barrier. ;)

But I think what the PR does now is good enough.